### PR TITLE
Re-enabled updating encrypted custom fields via API [sc-41465]

### PR DESCRIPTION
**Pull request by uberbrady**
Some recent tweaks to the API code made it so that updates to encrypted custom fields were not correctly being encrypted, and, thus, not decrypted either.

This fixes that, and adds new tests to cover those possibilities.

Additionally, we decided to add some new parts - such as when you _try_ to update a custom field, but fail, because you don't have "admin" permissions. Now we still allow the request to succeed, but add a warning in the message. Tests now cover this possibility.

The one weird thing here is that there were *TWO* return statements in the API - one that, obviously, fired. And one that was being ignored. I decided to just simply delete the second one, since it was not actually being run, ever.

I think there are a lot of inefficiencies with how I've done the testing, and will probably tap @marcusmoore for some help on more Laravel-esque ways of setting those test factories up in a bit of a less-janky fashion going forward. But I'd still go with this for now, as it seems like it's more in the right direction.

### CODE REVIEW

File: tests/Feature/Api/Assets/AssetStoreTest.php
Comment (marcusmoore) [inline tests/Feature/Api/Assets/AssetStoreTest.php]: Since we're sending `patch` requests we should move this to a new `AssetUpdateTest` in the same namespace.

Maybe a new version of this test can be `testEncryptedCustomFieldCanBeStored`.

File: tests/Feature/Api/Assets/AssetStoreTest.php
Comment (marcusmoore) [inline tests/Feature/Api/Assets/AssetStoreTest.php]: I would split these into two separate tests.

Maybe `testEncryptedCustomFieldCanBeUpdated` and `testPermissionNeededToUpdateEncryptedField`.

File: database/factories/AssetModelFactory.php
Comment (marcusmoore) [inline database/factories/AssetModelFactory.php]: I would change this to `hasEncryptedField()` or `withEncryptedField()`. I don't know if I have a preferences aside from the usage reading more like a sentence: `AssetModel::factory()->withEncryptedField()->create()`.

File: tests/Support/CustomTestMacros.php
Comment (marcusmoore) [inline tests/Support/CustomTestMacros.php:92]: Nice helper addition 😄

File: database/factories/AssetFactory.php
Comment (marcusmoore) [inline database/factories/AssetFactory.php]: Can we make this `hasEncryptedCustomField()` to match casing?

File: database/factories/AssetModelFactory.php
Comment (marcusmoore) [inline database/factories/AssetModelFactory.php]: I'm going to see if I can help make these a little clearer.

**Review by marcusmoore — CHANGES_REQUESTED** [review]
I added some general comments and I'm going to dive deeper into the factories.

**Comment by uberbrady** [comment]
@marcusmoore I think I made just about all of the choices you asked for - feel free to take another look, or if you want to 'fix' the factory thing or point me the right way to do that - that's fine too.

Also, if you think we should merge this as-is and do a Shortcut writeup for the improvements, we can do that too.

**Comment by marcusmoore** [comment]
@uberbrady I'd rather have the factories tightened up before merging. I'll work on that.

I'll also see what's up with the MySQL test failures.

File: tests/Feature/Api/Assets/AssetStoreTest.php
Comment (marcusmoore) [reply tests/Feature/Api/Assets/AssetStoreTest.php]: This should be split into two separate tests like we did with `AssetUpdateTest` but splitting them makes the second assertion fail 🙃

(Generally running multiple requests in one test is a bad idea and can introduce some quirks.

**Comment by marcusmoore** [comment]
@uberbrady I totally forgot I've been down this path with testing custom fields before. MySQL tests are failing because the creation of a custom field runs a migration and Laravel's tests are being run in a transaction which doesn't like that...

We could do some toggling where tests for custom fields are only run on the sqlite driver but I really don’t want to segment our tests like that.

File: tests/Feature/Api/Assets/AssetStoreTest.php
Comment (uberbrady) [inline tests/Feature/Api/Assets/AssetStoreTest.php:520]: This Patch -> Post thing, is that important? Or is it pretty much the same thing? Or are we talking about, maybe, two different things - updating an existing thing versus making a new thing?

Comment (marcusmoore) [reply tests/Feature/Api/Assets/AssetStoreTest.php:520]: > Or are we talking about, maybe, two different things - updating an existing thing versus making a new thing?

Yeah that's pretty much it. Our docs (and http standards I think) say [`POST` for creating a new thing](https://snipe-it.readme.io/reference/hardware-create) and [`PATCH` for partially updating a thing](https://snipe-it.readme.io/reference/hardware-partial-update). Side-note: we also support [PUT for updating a thing fully](https://snipe-it.readme.io/reference/hardware-update) (meaning all properties are present in the request).

**Comment by snipe** [comment]
Tests seem to be failing on this @marcusmoore @uberbrady

**Comment by marcusmoore** [comment]
@snipe Yeah...I don't think we'll be able to get around that unfortunately (see my last comment) 😞 

I hate to say it but it's probably best to remove the test cases from this PR 🙁

**Comment by uberbrady** [comment]
We talked internally about possible ways of not pulling the tests entirely, and the change that I pushed up later adds some code to mark a test as 'incomplete' if the test database driver is 'mysql' - so you can still run the tests locally against sqlite, but they won't completely explode on a GH action in a PR on the 'server side'.

While not ideal (you don't want to ever have to say "if this driver is mysql..." in a test), it's at least half-there, and maybe at some point in the future we can get all-there.

@marcusmoore was able to clean up my fixes into something a little more systemic, which was good!

**Comment by marcusmoore** [comment]
Gonna give this another review pass. There is one todo I added that I might be able to easily handle.

**Comment by marcusmoore** [comment]
Note: I think it'll be a good idea to review/merge #14458 and then apply this change on top of it. Looking at the changed lines in the controllers I think the merge conflicts will be easier to handle in this PR than that one.

**Review by marcusmoore — APPROVED** [review]
I removed my todo since it wasn't holding up this PR.

This looks good but remember that it'll probably be easier to get #14458 in first and then add this PR on top of it.

**Pull request merged**
